---
title: "Evolutionary dynamics of ncORFs"
editor: source
editor_options: 
  chunk_output_type: console
---

## GnomAD v3 constraint Z score (Gnocchi score)
The flagship paper calculated constraint Z-score (CZS; Gnocchi in the final published version) for 1000-bp sliding windows with a step size of 100bp.
Based on this, we calculate CZS for each 100-bp window by average CZS of all overlapping 1000-bp windows.
For each ORF, we calculate average CZS by weighted average of CZS of windows that overlapped this ORF.
For example: 3 bases in w1, 6 bases in w2. scores = 1/3 * Z(w1) + 2/3 * Z(w2)

note: Gnocchi score is calculated using gnomAD v3.1.2 (https://github.com/atgu/gnomad_nc_constraint)

```{bash}
# CZS: /nfs_data/database/gnomAD/flagship_paper_v3/sd3_constraintZscore_w1kb_s100bp_grch38.bed.gz
# pwd: /nfs_data/zhangh/ncorf_mammals/variation
grep '^[0-9]' /nfs_data/database/ref_genomes/human_GRCh38p13/ens107/genome.fa.fai | sed 's/^/chr/' >hg38.sizes
bedtools makewindows -g hg38.sizes -w 100 -s 100 >hg38.100bp.windows.bed
zcat /nfs_data/database/gnomAD/flagship_paper_v3/sd3_constraintZscore_w1kb_s100bp_grch38.bed.gz |  bedtools map -a hg38.100bp.windows.bed -b - -c 4 -o mean,count | sed 's/^chr//' | gzip -c >hg38.100bp.win.czs.bed.gz

bedtools bed12tobed6 -i ../data/human_three_methods_ncorf.bed | bedtools sort | \
    bedtools intersect -wao -a - -b <(zcat hg38.100bp.win.czs.bed.gz) \
    >human_three_methods_ncorf.czs.bed

# also calculate CZS of CDS and ncORF for comparison
gppy convert2bed -t cds -g /nfs_data/database/ref_genomes/human_GRCh38p13/ens107/Homo_sapiens.GRCh38.107.gtf.gz >Homo_sapiens.GRCh38.107.cds.bed
bedtools bed12tobed6 -i Homo_sapiens.GRCh38.107.cds.bed | bedtools sort | \
    bedtools intersect -wao -a - -b <(zcat hg38.100bp.win.czs.bed.gz) \
    >Homo_sapiens.GRCh38.107.cds.czs.bed
    
bedtools bed12tobed6 -i human_three_methods_ncorf.bed | bedtools sort | 
    bedtools intersect -wao -a - -b <(zcat /nfs_data/zhangh/ncorf_mammals/variation/hg38.100bp.win.czs.bed.gz) \
    >human_three_methods_ncorf.czs.bed
```

## gnocchi score
```{r}
library(data.table)
library(ggplot2)
library(cols4all)

theme_set(theme_classic(base_size = 12) + theme(
    axis.text = element_text(color = 'black'),
    strip.background = element_blank(),
    strip.text = element_text(size = 12),
    plot.title = element_text(hjust = 0.5),
    plot.subtitle = element_text(hjust = 0.5)))

# orfs
czs <- fread('data/variation/human_three_methods_ncorf.czs.bed', header = FALSE)
czs[, V10 := as.numeric(V10)]
czs_orf <- czs[V11 > 0, .(len_covered = sum(V12), mean_czs = sum(V10 * V12/sum(V12))), by = .(orf_id = V4)]

czs_orf_all <- unique(czs[, .(orf_id = V4)])
czs_orf_all <- czs_orf[czs_orf_all, on = .(orf_id)]
fwrite(czs_orf_all, 'data/variation/human_three_methods_ncorf.czs.tsv', sep = '\t')

# cds
czs2 <- fread('data/variation/Homo_sapiens.GRCh38.107.cds.czs.bed', header = FALSE)
czs2[, V10 := as.numeric(V10)]
czs_cds <- czs2[V11 > 0, .(len_covered = sum(V12), mean_czs = sum(V10 * V12/sum(V12))), by = .(tx_name = V4)]

czs_cds_all <- unique(czs2[, .(tx_name = V4)])
czs_cds_all <- czs_cds[czs_cds_all, on = .(tx_name)]
fwrite(czs_orf_all, 'data/variation/human_three_methods_final_ncorfs.czs.tsv', sep = '\t')


# compare
ncorfs_cdsovlp <- fread('data/ncorf_annot/human_three_methods_ncorf.cds_overlap.tsv', header = FALSE)
ncorfs_cdsovlp <- ncorfs_cdsovlp[V13 > 0]
txrep_hsa_pcg <- fread('data/Homo_sapiens.GRCh38.107.txinfo.rep.pcg.tsv')

plt <- rbindlist(list(
    ncORF = czs_orf_all[!is.na(mean_czs) & !orf_id %in% ncorfs_cdsovlp$V4, .(id = orf_id, gnocchi = mean_czs)],
    CDS = czs_cds_all[!is.na(mean_czs) & tx_name %in% txrep_hsa_pcg$tx_name, .(id = tx_name, gnocchi = mean_czs)]
), idcol = 'type')

ggplot(plt, aes(x = gnocchi, fill = type)) +
    geom_histogram(position = 'identity', alpha = 0.618, bins = 50) +
    geom_vline(xintercept = 0, linetype = 2) +
    labs(x = 'Gnocchi score', y = 'Count', fill = NULL) +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_fill_discrete_c4a_cat('carto.vivid') +
    theme(legend.position = c(0.02, 0.98),
          legend.justification = c(0, 1))
ggsave('figures/ncorf_cds_gnocchi_comparison.pdf', width = 4, height = 3)

ggplot(plt, aes(x = gnocchi, fill = type)) +
    geom_density(position = 'identity', alpha = 0.618) +
    geom_vline(xintercept = 0, linetype = 2) +
    labs(x = 'Gnocchi score', y = 'Count', fill = NULL) +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    scale_fill_discrete_c4a_cat('carto.vivid') +
    theme(legend.position = c(0.02, 0.98),
          legend.justification = c(0, 1))
ggsave('figures/ncorf_cds_gnocchi_comparison_density.pdf', width = 4, height = 3)


t.test(plt[type == 'CDS', gnocchi], mu = 0, alternative = 'greater')$p.value  # 0
t.test(plt[type == 'ncORF', gnocchi], mu = 0, alternative = 'greater')$p.value # 0
plt[, .(mean = mean(gnocchi), median = median(gnocchi), prop = sum(gnocchi > 0)/.N), by = .(type)]
#     type     mean   median      prop
# 1: ncORF 1.331047 1.382157 0.7689735
# 2:   CDS 1.873174 1.862607 0.8856956
```

# datasets (mammals or primates)
## PhyloP
prepare:
- get intervals that are expanded by 30bp at both 5' end and 3' end with `bed12_slop.py`;
- extract per-base phylop with `bigwig_covpn_dna.py`;
- summarize in R;

```{bash}
# ncORFs
python /nfs_data/zhangh/tmp/bed12_slop.py -i ../data/human_three_methods_final_ncorfs.bed -l -30 -r -30 -o human_three_methods_final_ncorfs_30bp.bed
sed 's/^/chr/g' human_three_methods_final_ncorfs_30bp.bed >human_three_methods_final_ncorfs_30bp_chr.bed

python bigwig_covpn_dna.py hg38.phyloP447way_primate.bw human_three_methods_final_ncorfs_30bp_chr.bed >hg38.primate_phylop_ncorf.tsv
python bigwig_covpn_dna.py hg38.phyloP447way_mammals.bw human_three_methods_final_ncorfs_30bp_chr.bed >hg38.mammals_phylop_ncorf.tsv

# CDSs
gppy convert2bed -g ../Homo_sapiens.GRCh38.107.gtf -t cds -e 30 >Homo_sapiens.GRCh38.107.bed
sed 's/^/chr/g' Homo_sapiens.GRCh38.107.cds.bed >Homo_sapiens.GRCh38.107.cds_chr.bed

python bigwig_covpn_dna.py hg38.phyloP447way_primate.bw Homo_sapiens.GRCh38.107.cds_chr.bed >Homo_sapiens.GRCh38.107.primate_phylop_cds.tsv
python bigwig_covpn_dna.py hg38.phyloP447way_mammals.bw Homo_sapiens.GRCh38.107.cds.30bp.chr.bed >Homo_sapiens.GRCh38.107.mammals_phylop_cds.tsv
```

compress phyloP data on disk.
```{bash}
# /nfs_data/zhangh/ncorf_mammals/evolution
# copied from the following locations
cp /nfs_data/changy/poj/ORF/human_riboseq/phylop/hg38.phyloP447way.mammals.ncorf.tsv.gz \
/nfs_data/changy/poj/ORF/human_riboseq/phylop/hg38.phyloP447way.mammals.cds.tsv.gz \
/nfs_data/changy/poj/ORF/human_riboseq/phylop/hg38.phyloP447way.primate.ncorf.tsv.gz \
/nfs_data/changy/poj/ORF/human_riboseq/phylop/hg38.phyloP447way.primate.cds.tsv.gz .
```

```{r}
library(data.table)
library(ggplot2)
library(patchwork)
library(cols4all)

theme_set(theme_classic(base_size = 12) + theme(
    axis.text = element_text(color = 'black'),
    strip.background = element_blank(),
    strip.text = element_text(size = 12),
    plot.title = element_text(hjust = 0.5),
    plot.subtitle = element_text(hjust = 0.5)))
```

check name collisions between human and mouse ncORFs
```{r}
ncorfs_human <- fread('data/ncorf_annot/human_three_methods_ncorf.final.tsv')
ncorfs_human[, orf_id := uid]

ncorfs_mouse <- fread('data/ncorf_annot/mouse_three_methods_ncorf.final.tsv')
ncorfs_mouse[, orf_id := uid]

intersect(ncorfs_human$orf_id, ncorfs_mouse$orf_id)
# character(0) -> empty. we can safely combine files from the two species

# human ncORFs overlapping with CDS
ncorfs_cdsovlp <- fread('data/ncorf_annot/human_three_methods_ncorf.cds_overlap.tsv', header = FALSE)
ncorfs_cdsovlp <- ncorfs_cdsovlp[V13 > 0]

# representative transcripts in human
txrep_hsa_pcg <- fread('data/Homo_sapiens.GRCh38.107.txinfo.rep.pcg.tsv')
```

## PhyloP
```{r}
nt_cutoff <- 30

read_phylop <- function(path){
    tab <- fread(path, sep = '\t', header = FALSE)
    tab <- tab[, .(id = V4, len_region = V7,
                   phylop = lapply(strsplit(V8, ',', fixed = TRUE), as.numeric))]
    return(tab)
}

start_phylop <- function(tab){
    tab <- tab[len_region >= (nt_cutoff + 60)]
    mat <- sapply(tab$phylop, \(x) x[1:(nt_cutoff + 30)])
    mat <- t(mat)
    rownames(mat) <- tab$id
    return(mat)
}

stop_phylop <- function(tab){
    tab <- tab[len_region >= (nt_cutoff + 60)]
    mat <- sapply(tab$phylop, tail, nt_cutoff + 30)
    mat <- t(mat)
    rownames(mat) <- tab$id
    return(mat)
}

phylop_mammals_ncorfs <- read_phylop('data/evolution/hg38.phyloP447way.mammals.ncorf.tsv.gz')
phylop_mammals_ncorfs <- phylop_mammals_ncorfs[!id %in% ncorfs_cdsovlp$V4]

phylop_primates_ncorfs <- read_phylop('data/evolution/hg38.phyloP447way.primate.ncorf.tsv.gz')
phylop_primates_ncorfs <- phylop_primates_ncorfs[!id %in% ncorfs_cdsovlp$V4]


start_mammals_ncorfs <- start_phylop(phylop_mammals_ncorfs)
start_primates_ncorfs <- start_phylop(phylop_primates_ncorfs)
stop_mammals_ncorfs <- stop_phylop(phylop_mammals_ncorfs)
stop_primates_ncorfs <- stop_phylop(phylop_primates_ncorfs)

# plot(colMeans(start_mammals_ncorfs), type = 'h')
# plot(apply(start_mammals_ncorfs, 2, median), type = 'h')

ptab_ncorfs <- rbindlist(list(
    start_Mammals = data.table(
        posn = 1:(nt_cutoff + 30) - 31,
        mean = colMeans(start_mammals_ncorfs),
        median = apply(start_mammals_ncorfs, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    start_Primates = data.table(
        posn = 1:(nt_cutoff + 30) - 31,
        mean = colMeans(start_primates_ncorfs),
        median = apply(start_primates_ncorfs, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    stop_Mammals = data.table(
        posn = 1:(nt_cutoff + 30) - nt_cutoff,
        mean = colMeans(stop_mammals_ncorfs),
        median = apply(stop_mammals_ncorfs, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    stop_Primates = data.table(
        posn = 1:(nt_cutoff + 30) - nt_cutoff,
        mean = colMeans(stop_primates_ncorfs),
        median = apply(stop_primates_ncorfs, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    )
), idcol = 'region') 


phylop_mammals_cds <- read_phylop('data/evolution/hg38.phyloP447way.mammals.cds.tsv.gz')
phylop_mammals_cds <- phylop_mammals_cds[id %in% txrep_hsa_pcg$tx_name]

phylop_primates_cds <- read_phylop('data/evolution/hg38.phyloP447way.primate.cds.tsv.gz')
phylop_primates_cds <- phylop_primates_cds[id %in% txrep_hsa_pcg$tx_name]


start_mammals_cds <- start_phylop(phylop_mammals_cds)
start_primates_cds <- start_phylop(phylop_primates_cds)
stop_mammals_cds <- stop_phylop(phylop_mammals_cds)
stop_primates_cds <- stop_phylop(phylop_primates_cds)

# plot(colMeans(start_mammals_cds), type = 'h')
# plot(apply(start_mammals_cds, 2, median), type = 'h')

ptab_cds <- rbindlist(list(
    start_Mammals = data.table(
        posn = 1:(nt_cutoff + 30) - 31,
        mean = colMeans(start_mammals_cds),
        median = apply(start_mammals_cds, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    start_Primates = data.table(
        posn = 1:(nt_cutoff + 30) - 31,
        mean = colMeans(start_primates_cds),
        median = apply(start_primates_cds, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    stop_Mammals = data.table(
        posn = 1:(nt_cutoff + 30) - nt_cutoff,
        mean = colMeans(stop_mammals_cds),
        median = apply(stop_mammals_cds, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    ),
    stop_Primates = data.table(
        posn = 1:(nt_cutoff + 30) - nt_cutoff,
        mean = colMeans(stop_primates_cds),
        median = apply(stop_primates_cds, 2, median),
        group = c('a', 'b')[(seq_len(nt_cutoff + 30) -1) %/% 3 %% 2 + 1]
    )
), idcol = 'region')

ptab <- rbind(CDS = ptab_cds, ncORF = ptab_ncorfs, idcol = 'type')
ptab[, c('region', 'taxa') := tstrsplit(region, '_')]
ptab <- ptab[, .(region, taxa, type, posn, mean, median, group)]

ggplot(ptab[region == 'start'], aes(x = posn, y = mean, fill = group)) +
    geom_col(show.legend = FALSE, width = 0.618) +
    facet_wrap(vars(taxa, type), scales = 'free', ncol = 1,
               labeller = label_wrap_gen(multi_line = FALSE)) +
    scale_fill_brewer(palette = 'Set1')
ggsave('figures/phylop_447way_start_mean.pdf', width = 12, height = 8)

ggplot(ptab[region == 'stop'], aes(x = posn, y = mean, fill = group)) +
    geom_col(show.legend = FALSE, width = 0.618) +
    facet_wrap(vars(taxa, type), scales = 'free', ncol = 1,
               labeller = label_wrap_gen(multi_line = FALSE)) +
    scale_fill_brewer(palette = 'Set1')
ggsave('figures/phylop_447way_stop_mean.pdf', width = 12, height = 8)

ggplot(ptab[taxa == 'Mammals'], aes(x = posn, y = mean, fill = factor(seq_along(group) %% 3))) +
    geom_col(show.legend = FALSE, width = 0.618) +
    scale_x_continuous(expand = expansion(mult = c(0.01, 0.01))) +
    facet_grid(rows = vars(type), cols = vars(region), scales = 'free') +
    scale_fill_discrete_c4a_cat('carto.safe') +
    labs(x = 'Position', y = 'PhyloP')
ggsave('figures/phylop_447way_mammal_mean.pdf', width = 12, height = 4)

ggplot(ptab[taxa == 'Primates'], aes(x = posn, y = mean, fill = factor(seq_along(group) %% 3))) +
    geom_col(show.legend = FALSE, width = 0.618) +
    facet_grid(rows = vars(type), cols = vars(region), scales = 'free') +
    scale_fill_discrete_c4a_cat('carto.safe') +
    labs(x = 'Position', y = 'PhyloP')
ggsave('figures/phylop_447way_primate_mean.pdf', width = 12, height = 4)
```

Significance test of periodicity based on autocorrelation
```{r}
library(testcorr)

# view autocorrelation
ptab[region == 'start' & taxa == 'Mammals' & type == 'CDS' & posn >= 3, acf(median)]
ptab[region == 'start' & taxa == 'Mammals' & type == 'CDS' & posn >= 3, ac.test(median, max.lag = 6)]

#' testing $rho_{lag=3} = 0$ should be done with t-type statistics or
#' its robust version according to manual. Here we choose the robust version (ttilde and pvttilde)
ptab_actest_start <- ptab[region == 'start' & posn >= 3, .(
    p_mean = ac.test(mean, max.lag = 3, plot = FALSE)$pvttilde[3],
    p_median = ac.test(median, max.lag = 3, plot = FALSE)$pvttilde[3]
), by = .(taxa, type)]

ptab_actest_stop <- ptab[region == 'stop' & posn <= -3, .(
    p_mean = ac.test(mean, max.lag = 3, plot = FALSE)$pvttilde[3],
    p_median = ac.test(median, max.lag = 3, plot = FALSE)$pvttilde[3]
), by = .(taxa, type)]

ptab_actest <- rbind(start = ptab_actest_start, stop = ptab_actest_stop, idcol = 'region')
ptab_actest
fwrite(ptab_actest, 'results/phylop_autocorrelation_test_30nt.tsv', sep='\t')

rm(phylop_mammals_cds, phylop_primates_cds, phylop_mammals_ncorfs, phylop_primates_ncorfs)
```

## ncORF origin
- view correspondance between node number and label: `View(as_tibble(tree_human))`
origin node infer methods
- naive: the most recent common ancestor of leaves with ORFs
- origin: based on ancestral sequence reconstruction. The most distant human ancestor between human
          and which all intermediate nodes have intact ORFs.

### reconstructed ancestral sequence
```{bash}
## human
# get fasta
python3 ~admin/scripts/maf_getfasta.py /nfs_data/database/hg38_100way/maf hg38 ../data_174/human_three_methods_ncorf_chr.bed output.maf 2>maf_getfasta.log
python ~admin/scripts/maf_splitfa.py output.maf maf_block/
# filter length
for i in maf_block/*; do j=$(basename $i); echo Rscript filter_seq_length.R $i mafft_block/${j%%.fa}.filter.fa; done >01_filter_length.sh
# mafft
for i in mafft_block/*.filter.fa; do j=$(basename $i); echo mafft-linsi $i \>mafft_block/${j%%.fa}.mafft.fa; done>02_mafft.sh
for i in mafft_block/*.mafft.fa; do j=$(basename $i); echo mkdir reconstruction_asr/${j%%.filter.mafft.fa}; done>03_mkdir_mafft.sh
for i in mafft_block/*.mafft.fa; do j=$(basename $i); echo mv $i reconstruction_asr/${j%%.filter.mafft.fa};done >04_mv_mafft.sh
# gotree
for i in reconstruction_asr/*/*.mafft.fa; do echo "grep '>' $i >${i%%.mafft.fa}_tips.txt"; done >05_get_tips.sh
for i in reconstruction_asr/*/*_tips.txt; do echo "sed 's/>//g' $i > ${i%%_tips.txt}_tips_no.txt"; done > 06_remove_tips.sh
for i in reconstruction_asr/*/*tips_no.txt; do echo gotree prune -i /nfs_data/database/hg38_100way/hg38.100way.nh -r true -f $i -o ${i%%_tips_no.txt}.nwk; done >07_gotree.sh
# fastml (SubMatrix HKY/JC_Nuc --OptimizeBL no --indelReconstruction ML)
for i in reconstruction_asr/*/*.mafft.fa; do echo perl /home/admin/local/FastML.v3.11/www/fastml/FastML_Wrapper.pl --MSA_file $i --Tree ${i%%.filter.mafft.fa}.nwk --seqType nuc --SubMatrix HKY --OptimizeBL no --indelReconstruction ML --outDir /nfs_data/changy/poj/ORF/human_riboseq/evolution/${i%/*.filter.mafft.fa}; done >08_fastml.sh
for i in tmp/*; do j=$(basename $i); echo perl /home/admin/local/FastML.v3.11/www/fastml/FastML_Wrapper.pl --MSA_file reconstruction_asr/$j/$j*.mafft.fa --Tree reconstruction_asr/$j/$j.nwk --seqType nuc --SubMatrix JC_Nuc --OptimizeBL no --indelReconstruction ML --outDir /nfs_data/changy/poj/ORF/human_riboseq/evolution_174/reconstruction_asr/$j; done >08_fastml_jcnuc.sh
for i in reconstruction_asr/*/*.mafft.fa; do echo mv ${i%/*.mafft.fa}/tree.newick.txt ${i%%.filter.mafft.fa}.tree.newick.txt ;done >09_mv_nwk.sh
for i in tmp/*; do j=$(basename $i); echo mv reconstruction_asr/$j/tree.newick.txt reconstruction_asr/$j/$j.tree.newick.txt ;done >09_mv_nwk_jcnuc.sh
for i in reconstruction_asr/*/*.mafft.fa; do echo mv ${i%/*.mafft.fa}/seq.marginal_IndelAndChars.txt ${i%%.filter.mafft.fa}.seq.marginal_IndelAndChars.txt ;done >10_mv_seq.sh
sed 's/_IndelAndChars//g' 10_mv_seq.failed.sh >10_mv_seq_noindel.sh

## mouse
# maf index
for i in *.maf; do echo python /home/admin/scripts/maf_index.py $i mm39.${i%%.maf} ;done > 00_index_maf.sh
# get fasta
python3 ~admin/scripts/maf_getfasta.py /nfs_data/database/mm39_35way/maf mm39 ../data_209/mouse_three_methods_ncorf_chr.bed output.maf 2>maf_getfasta.log
python ~admin/scripts/maf_splitfa.py output.maf maf_block/
# filter length
for i in maf_block/*; do j=$(basename $i); echo Rscript filter_seq_length.R $i mafft_block/${j%%.fa}.filter.fa; done >01_filter_length.sh
# mafft
for i in mafft_block/*.filter.fa; do j=$(basename $i); echo mafft-linsi $i \>mafft_block/${j%%.fa}.mafft.fa; done>02_mafft.sh
for i in mafft_block/*.mafft.fa; do j=$(basename $i); echo mkdir reconstruction_asr/${j%%.filter.mafft.fa}; done>03_mkdir_mafft.sh
for i in mafft_block/*.mafft.fa; do j=$(basename $i); echo mv $i reconstruction_asr/${j%%.filter.mafft.fa};done >04_mv_mafft.sh
# gotree (GCF_003668045 == CriGri)
for i in reconstruction_asr/*/*.mafft.fa; do echo "grep '>' $i >${i%%.mafft.fa}_tips.txt"; done >05_get_tips.sh
for i in reconstruction_asr/*/*_tips.txt; do echo "sed 's/>//g' $i > ${i%%_tips.txt}_tips_no.txt"; done > 06_remove_tips.sh
for i in reconstruction_asr/*/*tips_no.txt; do echo gotree prune -i mm39.35way.nh -r true -f $i -o ${i%%_tips_no.txt}.nwk; done >07_gotree.sh
# fastml
for i in reconstruction_asr/*/*.mafft.fa; do echo perl /home/admin/local/FastML.v3.11/www/fastml/FastML_Wrapper.pl --MSA_file $i --Tree ${i%%.filter.mafft.fa}.nwk --seqType nuc --SubMatrix HKY --OptimizeBL no --indelReconstruction ML --outDir /nfs_data/changy/poj/ORF/mouse_riboseq/evolution_209/${i%/*.filter.mafft.fa}; done >08_fastml.sh
for i in tmp/*; do j=$(basename $i); echo perl /home/admin/local/FastML.v3.11/www/fastml/FastML_Wrapper.pl --MSA_file reconstruction_asr/$j/$j*.mafft.fa --Tree reconstruction_asr/$j/$j.nwk --seqType nuc --SubMatrix JC_Nuc --OptimizeBL no --indelReconstruction ML --outDir /nfs_data/changy/poj/ORF/mouse_riboseq/evolution_209/reconstruction_asr/$j; done >08_fastml_jcnuc.sh
for i in reconstruction_asr/*/*.mafft.fa; do echo mv ${i%/*.mafft.fa}/tree.newick.txt ${i%%.filter.mafft.fa}.tree.newick.txt ;done >09_mv_nwk.sh
for i in tmp/*; do j=$(basename $i); echo mv reconstruction_asr/$j/tree.newick.txt reconstruction_asr/$j/$j.tree.newick.txt ;done >09_mv_nwk_jcnuc.sh
for i in reconstruction_asr/*/*.mafft.fa; do echo mv ${i%/*.mafft.fa}/seq.marginal_IndelAndChars.txt ${i%%.filter.mafft.fa}.seq.marginal_IndelAndChars.txt ;done >10_mv_seq.sh
sed 's/_IndelAndChars//g' 10_mv_seq.failed.sh >10_mv_seq_noindel.sh
```

## BLS
  - human：/nfs_data/changy/poj/ORF/human_riboseq/evolution_174/human_bls_ncorf.txt
  - mouse: /nfs_data/changy/poj/ORF/mouse_riboseq/evolution_209/mouse_bls_ncorf.txt
  
```{bash}
# human
for i in reconstruction_asr/*/*seq.marginal_IndelAndChars.txt; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_bls.py $i ${i%%.seq.marginal_IndelAndChars.txt}.tree.newick.txt hg38.100way.nh hg38 -o bls_result/${j%%.seq.marginal_IndelAndChars.txt}; done >11_human_bls.sh
for i in human_noindels/*; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_bls.py reconstruction_asr/$j/$j.seq.marginal.txt reconstruction_asr/$j/$j.tree.newick.txt hg38.100way.nh hg38 -o bls_result/$j; done >11_human_bls_noindels.sh

# mouse
for i in reconstruction_asr/*/*seq.marginal_IndelAndChars.txt; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_bls.py $i ${i%%.seq.marginal_IndelAndChars.txt}.tree.newick.txt mm39.35way.root.nh mm39 -o bls_result/${j%%.seq.marginal_IndelAndChars.txt}; done >11_mouse_bls.sh
for i in mouse_noindels/*; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_bls.py reconstruction_asr/$j/$j.seq.marginal.txt reconstruction_asr/$j/$j.tree.newick.txt mm39.35way.root.nh mm39 -o bls_result/$j; done >11_mouse_bls_noindels.sh
```

## BLS
```{r}
# merge bls
files_human <- list.files('evolution_174/bls_result/', pattern = 'result.txt' , full.names = T)
names(files_human) <- sub('bls_result/.*', '', basename(files_human))
sum_data_human <- bind_rows(lapply(files_human, read.table), .id = 'id') %>%   
  mutate(id = gsub('\\.result.txt', '', id))
write_tsv(sum_data_human, "human_bls_result.txt")

files_mouse <- list.files('evolution_174/bls_result/', pattern = 'result.txt' , full.names = T)
names(files_mouse) <- sub('bls_result/.*', '', basename(files_mouse))
sum_data_mouse <- bind_rows(lapply(files_mouse, read.table), .id = 'id') %>%   
  mutate(id = gsub('\\.result.txt', '', id))
write_tsv(sum_data, "mouse_bls_result.txt")

read_bls <- function(path){
    bls <- fread(path)
    bls <- bls[, seq_len(ncol(bls)) %% 2 == 1, with = FALSE]
    setnames(bls, c(
        "uid", "origin_manner", "origin_age", "origin_age_global", "bl_all", 
        "bl_sub_origin", "bl_orf_origin", "bls_global_origin", "bls_local_origin", 
        "naive_age_global", "bl_sub_naive", "bl_orf_naive", "bls_global_naive", 
        "bls_local_naive"))
    return(bls)
}

bls_human <- read_bls('data/evolution/human_bls_ncorf.txt')
bls_human <- bls_human[uid %in% ncorfs_human$uid]
# some human specific ncORFs have no orthologs in other species
bls_human <- merge(ncorfs_human[, .(uid)], bls_human, all.x = TRUE)
bls_human[is.na(origin_manner), `:=`(
    origin_manner = 'nondenovo', origin_age = 'hg38', origin_age_global = 'hg38',
    bl_all = 18.46038, bl_sub_origin = 0, bl_orf_origin = 0, bls_global_origin = 0,
    bls_local_origin = 0, naive_age_global = 'hg38', bl_sub_naive = 0,
    bl_orf_naive = 0, bls_global_naive = 0, bls_local_naive = 0)]

bls_mouse <- read_bls('data/evolution/mouse_bls_ncorf.txt')
bls_mouse <- bls_mouse[uid %in% ncorfs_mouse$uid]
bls_mouse <- merge(ncorfs_mouse[, .(uid)], bls_mouse, all.x = TRUE)
bls_mouse[is.na(origin_manner), `:=`(
    origin_manner = 'nondenovo', origin_age = 'mm39', origin_age_global = 'mm39',
    bl_all = 8.778538, bl_sub_origin = 0, bl_orf_origin = 0, bls_global_origin = 0,
    bls_local_origin = 0, naive_age_global = 'mm39', bl_sub_naive = 0,
    bl_orf_naive = 0, bls_global_naive = 0, bls_local_naive = 0)]

ncorfs_bls <- rbind(Human = bls_human, Mouse = bls_mouse, idcol = 'species')

# check relationship between global age and local age
ggplot(ncorfs_bls, aes(x = bls_global_origin, y = bls_local_origin)) +
    geom_point() +
    geom_abline(slope = 1, intercept = 0, linetype = 2, color = 'red') +
    facet_wrap(vars(species))

plt <- melt(ncorfs_bls, id.vars = c('species', 'uid', 'origin_manner'),
     measure.vars = c('bls_global_origin', 'bls_local_origin',
                      'bls_global_naive', 'bls_local_naive'),
     value.name = 'bls')
plt[, c('tree', 'method') := tstrsplit(variable, '_')[2:3]]
plt[, variable := NULL]
plt[, tree := ifelse(tree == 'global', 'Global', 'Local')]
plt[, method := ifelse(method == 'origin', 'ASR', 'MRCA')]

ggplot(plt, aes(x = bls, fill = tree)) +
    geom_histogram(position = 'identity', alpha = 0.618, bins = 20) +
    facet_grid(rows = vars(species), cols = vars(method), scales = 'free') +
    scale_fill_brewer(palette = 'Dark2') +
    labs(x = 'Branch Length Score', y = 'Number of ORFs', fill = NULL)
ggsave('figures/ncORF_BLS_methods_comparison.pdf', width = 7, height = 5)

plt2 <- plt[method == 'ASR']
ggplot(plt2, aes(x = bls, fill = tree)) +
    geom_histogram(position = 'identity', alpha = 0.618, bins = 20) +
    facet_wrap(vars(species), scales = 'free') +
    scale_fill_brewer(palette = 'Dark2') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05))) +
    labs(x = 'Branch Length Score', y = 'Number of ORFs', fill = NULL)
ggsave('figures/ncORF_BLS_ASR_histogram.pdf', width = 7, height = 3)
```


Human
```{r}
# locally conserved (mrca or asr)
tmp <- bls_human[, .(
    uid, age_ASR = origin_age_global, bls_global_ASR = bls_global_origin, bls_local_ASR = bls_local_origin,
    age_MRCA = naive_age_global, bls_global_MRCA = bls_global_naive, bls_local_MRCA = bls_local_naive)]
lcon_human <- melt(tmp, id.vars = 'uid', measure.vars = patterns('age_', 'bls_global_', 'bls_local_'),
                   value.name = c('age', 'bls_global', 'bls_local'))
lcon_human[, method := ifelse(variable == '1', 'ASR', 'MRCA')]
lcon_human[, variable := NULL]


# tabulate ORF origination node
stat_age_human <- melt(bls_human, id.vars = 'uid',
                       measure.vars = c('naive_age_global', 'origin_age_global'))
stat_age_human[, method := ifelse(variable == 'naive_age_global', 'MRCA', 'ASR')]
stat_age_human <- stat_age_human[, .(cnt = .N), by = .(method, label = value)]
# treat those without alignments as human-specific
stat_age_human[label == 'hg38', cnt := cnt + nrow(ncorfs_human) - nrow(bls_human)]
stat_age_human <- dcast(stat_age_human, label ~ method)
# species specific
# hg38  212   68

stat_lcon_human <- lcon_human[bls_local > 0.9, .(cnt = .N), by = .(label = age, method)]
stat_lcon_human <- dcast(stat_lcon_human, label ~ method)
colSums(stat_lcon_human[, .(ASR, MRCA)])
#  ASR MRCA 
# 2037 1020


library(treeio)
library(ggtree)
library(tidytree)

# plot full human tree
tree_human <- read.tree('data/evolution/human_global_tree.nwk')
tree_human <- ape::rotateConstr(tree_human, rev(tree_human$tip.label))
tree_human$node.label[1] <- 'N1'
# use common name
ucsc_hg38 <- readLines('data/evolution/hg38.100way.commonNames.nh')
ucsc_hg38 <- stringr::str_extract(ucsc_hg38, "[ (]*([A-Za-z_']+):.*", group = 1)
ucsc_hg38[ucsc_hg38 == '_tropicalis'] <- 'Western clawed frog'
tree_human$tip.label <- gsub('_', ' ', ucsc_hg38)

ggtree(tree_human, branch.length = 'none') +
    geom_tiplab() +
    geom_nodelab(hjust = 0) +
    geom_rootedge(rootedge = 1) +
    coord_cartesian(clip = 'off') +
    theme(plot.margin = unit(c(1, 5, 1, 1), "lines"))
ggsave('figures/100way_tree_uncollapsed.pdf', height = 25, width = 10)

# plot collapsed tree with ORF counts
tdata_human <- data.table(label = tree_human$node.label)
tdata_human[stat_age_human, `:=`(ASR = i.ASR, MRCA = i.MRCA), on = .(label)]
tdata_human[stat_lcon_human, `:=`(lcon_ASR = i.ASR, lcon_MRCA = i.MRCA), on = .(label)]
tdata_human[, `:=`(ASR = ifelse(is.na(ASR), '', as.character(ASR)),
                   MRCA = ifelse(is.na(MRCA), '', as.character(MRCA)),
                   lcon_ASR = ifelse(is.na(lcon_ASR), '', as.character(lcon_ASR)),
                   lcon_MRCA = ifelse(is.na(lcon_MRCA), '', as.character(lcon_MRCA)))]


plot_tree_human <- function(tree_in, collapse_nodes=TRUE){
    p <- ggtree(tree_in, branch.length = 'none') +
        geom_tiplab() +
        geom_nodelab(aes(x = branch), hjust = 0.5, vjust=-0.2) +
        geom_rootedge(rootedge = 1) +
        coord_cartesian(clip = 'off') +
        theme(plot.margin = unit(c(1, 7, 1, 1), "lines"))
    if(collapse_nodes == FALSE) return(p)
    fill_col <- '#99d8c9'
    p2 <- p |> 
        scaleClade(119, .2) |> 
        scaleClade(122, .5) |> 
        scaleClade(123, .1) |> 
        scaleClade(135, .1) |> 
        scaleClade(159, .2) |> 
        scaleClade(165, .5) |> 
        scaleClade(167, .1) |> 
        scaleClade(186, .1) |> 
        collapse(119, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Old world monkeys') |>  # Cercopithecidae
        collapse(122, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'New world monkeys') |>  # Cebidae
        collapse(123, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Glires') |> 
        collapse(135, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Laurasiatheria') |>  
        collapse(159, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Afrotheria') |>
        collapse(165, mode = 'min', fill = fill_col, alpha = 0.6, 
                 clade_name = 'Marsupials') |>   
        collapse(167, mode = 'min', fill = fill_col, alpha = 0.6, 
                 clade_name = 'Reptilia & Aves') |>  
        collapse(186, mode = 'min', fill = fill_col, alpha = 0.6, 
                 clade_name = 'Actinopteri')  
    return(p2)
}

tree_tmp <- copy(tree_human)
tree_tmp$node.label <- tdata_human$ASR
p1 <- plot_tree_human(tree_tmp)
tree_tmp$node.label <- tdata_human$lcon_ASR
p2 <- plot_tree_human(tree_tmp)
p1 + p2
ggsave('figures/distribution_of_origin_human_asr.pdf', width = 12, height = 6)
```

Mouse
```{r}
# locally conserved (mrca or asr)
tmp <- bls_mouse[, .(
    uid, age_ASR = origin_age_global, bls_global_ASR = bls_global_origin, bls_local_ASR = bls_local_origin,
    age_MRCA = naive_age_global, bls_global_MRCA = bls_global_naive, bls_local_MRCA = bls_local_naive)]
lcon_mouse <- melt(tmp, id.vars = 'uid', measure.vars = patterns('age_', 'bls_global_', 'bls_local_'),
                   value.name = c('age', 'bls_global', 'bls_local'))
lcon_mouse[, method := ifelse(variable == '1', 'ASR', 'MRCA')]
lcon_mouse[, variable := NULL]

# tabulate ORF origination node
stat_age_mouse <- melt(bls_mouse, id.vars = 'uid',
                       measure.vars = c('naive_age_global', 'origin_age_global'))
stat_age_mouse[, method := ifelse(variable == 'naive_age_global', 'MRCA', 'ASR')]
stat_age_mouse <- stat_age_mouse[, .(cnt = .N), by = .(method, label = value)]
# treat those without alignments as mouse-specific
stat_age_mouse[label == 'mm39', cnt := cnt + nrow(ncorfs_mouse) - nrow(bls_mouse)]
stat_age_mouse <- dcast(stat_age_mouse, label ~ method)
# mm39 2465 1746

stat_lcon_mouse <- lcon_mouse[bls_local > 0.9, .(cnt = .N), by = .(label = age, method)]
stat_lcon_mouse <- dcast(stat_lcon_mouse, label ~ method)
colSums(stat_lcon_mouse[, .(ASR, MRCA)])
#  ASR MRCA 
# 4236 3343

# plot mouse full tree
tree_mouse <- read.tree('data/evolution/mouse_global_tree.nwk')
tree_mouse <- ape::rotateConstr(tree_mouse, rev(tree_mouse$tip.label))
tree_mouse$node.label[1] <- 'N1'
# use common name
ucsc_mm39 <- read.tree('data/evolution/mm39.35way.commonNames.nh')
tree_mouse$tip.label <- gsub('_', ' ', ucsc_mm39$tip.label)  
tree_mouse$tip.label[tree_mouse$tip.label == 'X. tropicalis'] <- 'Western clawed frog'

# View(as_tibble(tree_mouse))

ggtree(tree_mouse, branch.length = 'none', ladderize = FALSE) +
    geom_tiplab() +
    geom_nodelab(hjust = 0) +
    geom_rootedge(rootedge = 1) +
    coord_cartesian(clip = 'off') +
    theme(plot.margin = unit(c(1, 6, 1, 1), "lines"))
ggsave('figures/35way_tree_uncollapsed.pdf', height = 10, width = 8)

# plot collapsed tree with ORF counts
tdata_mouse <- data.table(label = tree_mouse$node.label)
tdata_mouse[stat_age_mouse, `:=`(ASR = i.ASR, MRCA = i.MRCA), on = .(label)]
tdata_mouse[stat_lcon_mouse, `:=`(lcon_ASR = i.ASR, lcon_MRCA = i.MRCA), on = .(label)]
tdata_mouse[, `:=`(ASR = ifelse(is.na(ASR), '', as.character(ASR)),
                   MRCA = ifelse(is.na(MRCA), '', as.character(MRCA)),
                   lcon_ASR = ifelse(is.na(lcon_ASR), '', as.character(lcon_ASR)),
                   lcon_MRCA = ifelse(is.na(lcon_MRCA), '', as.character(lcon_MRCA)))]

plot_tree_mouse <- function(tree_in, collapse_nodes=TRUE){
    p <- ggtree(tree_in, branch.length = 'none', ladderize = FALSE) +
        geom_tiplab() +
        geom_nodelab(aes(x = branch), hjust = 0.5, vjust=-0.2) +
        geom_rootedge(rootedge = 1) +
        coord_cartesian(clip = 'off') +
        theme(plot.margin = unit(c(1, 7, 1, 1), "lines"))
    
    if(collapse_nodes == FALSE) return(p)
    fill_col <- '#99d8c9'
    p2 <- p |> 
    scaleClade(50, .5) |> 
    scaleClade(52, .2) |> 
    scaleClade(60, .2) |> 
    scaleClade(69, .5) |> 
    collapse(50, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Rabbits') |>  # (Lagomorpha)
    collapse(52, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Primates') |> 
    collapse(60, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Laurasiatheria') |> 
    collapse(69, mode = 'min', fill = fill_col, alpha = 0.6,
                 clade_name = 'Afrotheria')
    return(p2)
}

tree_tmp <- copy(tree_mouse)
tree_tmp$node.label <- tdata_mouse$ASR
p1 <- plot_tree_mouse(tree_tmp)
tree_tmp$node.label <- tdata_mouse$lcon_ASR
p2 <- plot_tree_mouse(tree_tmp)
p1 + p2
ggsave('figures/distribution_of_origin_mouse_asr.pdf', width = 12, height = 6)


tree_tmp <- copy(tree_mouse)
tree_tmp$node.label <- tdata_mouse$MRCA
p1 <- plot_tree_mouse(tree_tmp)
tree_tmp$node.label <- tdata_mouse$lcon_MRCA
p2 <- plot_tree_mouse(tree_tmp)
p1 + p2
ggsave('figures/distribution_of_origin_mouse_mrca.pdf', width = 12, height = 6)
```

Relationship between age node and fraction of locally conserved ORFs
```{r}
tdata_human[setDT(as_tibble(tree_human)), `:=`(
    node = i.node, branch_length = i.branch.length), on = .(label)]
tdata_human[, dist2ref := castor::get_pairwise_distances(tree_human, rep(1, nrow(tdata_human)), node)]

tdata_mouse[setDT(as_tibble(tree_mouse)), `:=`(
    node = i.node, branch_length = i.branch.length), on = .(label)]
tdata_mouse[, dist2ref := castor::get_pairwise_distances(tree_mouse, rep(1, nrow(tdata_mouse)), node)]

origin_stat <- rbind(Human = tdata_human[node <= 118], Mouse = tdata_mouse[node <= 49], idcol = 'species')
origin_stat[, c('ASR', 'MRCA', 'lcon_ASR', 'lcon_MRCA') := lapply(.SD, as.numeric),
    .SDcols = c('ASR', 'MRCA', 'lcon_ASR', 'lcon_MRCA')]

ggplot(origin_stat, aes(x = dist2ref, y = lcon_ASR / ASR)) +
    geom_smooth(span = 1) +
    geom_point(size = 3, alpha = 0.6) +
    facet_wrap(vars(species), scales = 'free') +
    labs(x = 'Distance from origin node to tip',
         y = 'Fraction of locally conserved ORFs')
ggsave('figures/dist2ref_vs_locally_conserved_orfs_ASR.pdf', width = 6, height = 3)


ggplot(origin_stat, aes(x = dist2ref, y = lcon_MRCA / MRCA)) +
    geom_smooth(span = 1) +
    geom_point(size = 3, alpha = 0.6) +
    facet_wrap(vars(species), scales = 'free') +
    labs(x = 'Distance from origin node to tip',
         y = 'Fraction of locally conserved ORFs')
ggsave('figures/dist2ref_vs_locally_conserved_orfs_MRCA.pdf', width = 6, height = 3)


origin_stat[, .(
    ASR_corr = cor.test(dist2ref, lcon_ASR / ASR, method = 'spearman')$estimate,
    ASR_pval = cor.test(dist2ref, lcon_ASR / ASR, method = 'spearman')$p.value
), by = .(species)]
#    species   ASR_corr     ASR_pval
# 1:   Human -0.7647059 0.0003342936
# 2:   Mouse -0.1780220 0.5423155834

# ---------------------------------------------------------------------------------------------
# whether number of ORFs originated (scaled by branch length) correlate with node age?
origin_stat[, pseudo_bl := branch_length]
origin_stat[is.na(pseudo_bl), pseudo_bl := 1] 
# root node have no branch length, could be very large, but we only set to pseudo-value of 1

ggplot(origin_stat, aes(x = dist2ref, y = ASR/pseudo_bl)) +
    geom_smooth(method = 'lm') +
    geom_point(size = 3, alpha = 0.6) +
    scale_y_log10() +
    facet_wrap(vars(species), scales = 'free') +
    labs(x = 'Age of origination node', y = '# of ORFs / Branch length')
ggsave('figures/origin_age_vs_orf_numbers_ASR.pdf', width = 6, height = 3)


origin_stat[, .(
    ASR_corr = cor.test(dist2ref, ASR/pseudo_bl, method = 'spearman')$estimate,
    ASR_pval = cor.test(dist2ref, ASR/pseudo_bl, method = 'spearman')$p.value
), by = .(species)]
# "  species   ASR_corr    ASR_pval
# 1:   Human -0.6800826 0.002514381
# 2:   Mouse -0.7758242 0.001741787

# still significant after excluding the root node
origin_stat[, .(
    ASR_corr = cor.test(dist2ref, ASR/branch_length, method = 'spearman')$estimate,
    ASR_pval = cor.test(dist2ref, ASR/branch_length, method = 'spearman')$p.value
), by = .(species)]
#    species   ASR_corr    ASR_pval
#     <char>      <num>       <num>
# 1:   Human -0.6200980 0.009351552
# 2:   Mouse -0.7197802 0.007497812
```

save results
```{r}
ncorfs_cons[lcon_human, age := i.age, on = .(uid, method)]
ncorfs_cons[lcon_mouse, age := i.age, on = .(uid, method)]

fwrite(ncorfs_cons, 'results/ncorfs_bls_phylocsf.tsv', sep = '\t')
fwrite(origin_stat, 'results/ncorfs_origin_node_stat.tsv', sep = '\t')

ncorfs_cons[method == 'ASR', .(prop = sum(origin == 'denovo')/.N), by = .(species)]
# Key: <species>
#    species      prop
#     <char>     <num>
# 1:   Human 0.5070976
# 2:   Mouse 0.5053656
```

# conservation scores
## phyloCSF
  - human：/nfs_data/changy/poj/ORF/human_riboseq/evolution_174/human_phyloCSF_ncorf.txt
  - mouse: /nfs_data/changy/poj/ORF/mouse_riboseq/evolution_209/mouse_phyloCSF_ncorf.txt
  
```{bash}
# human
for i in bls_result/*result.txt; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_maf.py $i hg38.100way.nh maf_block/${j%%.result.txt}_filter_stop_codon.fa hg38 phylocsf_nostop_codon/${j%%.result.txt}_phylocsf.txt ; done>15_human_phylocsf.sh
for i in phylocsf_nostop_codon/*; do echo ~admin/miniforge3/bin/phylocsf++ score-msa --threads 16 100vertebrates $i ; done >16_human_score_msa.sh
# mouse
for i in bls_result/*result.txt; do j=$(basename $i); echo python ~zhangh/tmp/bls/orf_maf.py $i mm39.35way.root.nh maf_block/${j%%.result.txt}_filter_stop_codon.fa mm39 phylocsf_nostop_codon/${j%%.result.txt}_phylocsf.txt ; done>15_mouse_phylocsf.sh
for i in phylocsf_nostop_codon/*; do echo ~admin/miniforge3/bin/phylocsf++ score-msa --mapping mapping_name.tsv --threads 16 100vertebrates $i ; done >16_mouse_score_msa.sh
```

## PhyloCSF
```{r}
# rbind PhyloCSF
read_dtt <- function(x){
  dtt <- read.table(x, header = T)
}
files_human <- list.files('evolution_174/phylocsf_nostop_codon/', pattern = '_phylocsf.txt.scores' , full.names = T)
names(files_human) <- sub('phylocsf_nostop_codon/.*', '', basename(files_human))
sum_data_human <- bind_rows(lapply(files_human, read_dtt), .id = 'id') %>%   
  mutate(id = gsub('_phylocsf.txt.scores', '', id))
write_tsv(sum_data_human, "human_phyloCSF_ncorf.txt")

files_mouse <- list.files('evolution_209/phylocsf_nostop_codon/', pattern = '_phylocsf.txt.scores' , full.names = T)
names(files_mouse) <- sub('phylocsf_nostop_codon/.*', '', basename(files_mouse))
sum_data_mouse <- bind_rows(lapply(files_mouse, read_dtt), .id = 'id') %>%   
  mutate(id = gsub('_phylocsf.txt.scores', '', id))
write_tsv(sum_data_mouse, "mouse_phyloCSF_ncorf.txt")

read_phylocsf <- function(path){
    scores <- fread(path)
    scores <- scores[, .(uid = id, method = seq, orf_len = end,
                         phylocsf = phylocsf.score, codon_bls = bls.score)]
    scores[, method := ifelse(method == 'naive', 'MRCA', 'ASR')]
    return(scores)
}

phylocsf_human <- read_phylocsf('data/evolution/human_phyloCSF_ncorf.txt')
phylocsf_human <- phylocsf_human[uid %in% ncorfs_human$uid]

phylocsf_mouse <- read_phylocsf('data/evolution/mouse_phyloCSF_ncorf.txt')
phylocsf_mouse <- phylocsf_mouse[uid %in% ncorfs_mouse$uid]


# combine conservation scores: BLS & PhyloCSF
ncorfs_cons <- melt(
    ncorfs_bls,
    id.vars = c('species', 'uid', 'origin_manner'),
    measure.vars = c(
        'bls_global_origin', 'bls_local_origin',
        'bls_global_naive', 'bls_local_naive'),
    value.name = 'bls')
ncorfs_cons[, c('tree', 'method') := tstrsplit(variable, '_')[2:3]]
ncorfs_cons[, variable := NULL]
setnames(ncorfs_cons, 'origin_manner', 'origin')
ncorfs_cons <- dcast(ncorfs_cons, species + uid + origin + method ~ tree, value.var = 'bls')
setnames(ncorfs_cons, c('global', 'local'), c('bls_global', 'bls_local'))
ncorfs_cons[, method := ifelse(method == 'naive', 'MRCA', 'ASR')]  # all ncORFs are included

ncorfs_cons <- merge(ncorfs_cons, rbind(Human = phylocsf_human, Mouse = phylocsf_mouse, idcol='species'),
      by = c('species', 'uid', 'method'), all.x = TRUE)
ncorfs_cons[, phylocsf_pc := phylocsf/(orf_len/3)]  # get PhyloCSF per codon

p1 <- ggplot(ncorfs_cons[species == 'Human' & method == 'ASR'], aes(x = phylocsf_pc)) +
    geom_density(position = 'identity', show.legend = FALSE, fill = '#44AA99') +
    geom_vline(xintercept = 0, linetype = 2) +
    coord_cartesian(xlim = c(-70, 30)) +
    labs(x = 'PhyloCSF / number of codons', y = 'Density', title = 'Human') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))

p2 <- ggplot(ncorfs_cons[species == 'Mouse' & method == 'ASR'], aes(x = phylocsf_pc)) +
    geom_density(position = 'identity', show.legend = FALSE, fill = '#44AA99') +
    geom_vline(xintercept = 0, linetype = 2) +
    coord_cartesian(xlim = c(-20, 10)) +
    labs(x = 'PhyloCSF / number of codons', y = 'Density', title = 'Mouse') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))
p1 + p2
ggsave('figures/ncorf_phylocsf_density.pdf', width = 8, height = 3.5)


ggplot(ncorfs_cons[species == 'Human' & method == 'ASR'], aes(x = phylocsf_pc, y = bls_local)) +
    geom_point(alpha = 0.2) +
    geom_vline(xintercept = 0, linetype = 2) +
    coord_cartesian(xlim = c(-70, 30)) +
    labs(x = 'PhyloCSF per codon', y = 'Local BLS', title = 'Human') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))

ncorfs_cons[method == 'ASR', as.list(table(phylocsf_pc > 0, useNA = 'ifany')), by = .(species)]
# "  species FALSE  TRUE    NA
#     <char> <int> <int> <int>
# 1:   Human 10357  2103     9
# 2:   Mouse 11226  5555   179"

ncorfs_cons[method == 'ASR', as.list(table(phylocsf > log(1000), useNA = 'ifany')), by = .(species)]
# "  species FALSE  TRUE    NA
#     <char> <int> <int> <int>
# 1:   Human 10615  1845     9
# 2:   Mouse 12332  4449   179"

ncorfs_cons[method == 'ASR', as.list(table(bls_local > 0.9, useNA = 'ifany')), by = .(species)]
# "  species FALSE  TRUE
#     <char> <int> <int>
# 1:   Human 10432  2037
# 2:   Mouse 12724  4236"

# group by phylocsf > 0
plt <- copy(ncorfs_cons)
plt[, grp := ifelse(phylocsf > 0, 'PhyloCSF > 0', 'PhyloCSF < 0')]

ggplot(plt, aes(x = bls_local, fill = grp)) +
    geom_density(position = 'identity', alpha = 0.618) +
    facet_wrap(vars(species)) +
    labs(x = 'Local BLS', y = 'Density', fill = NULL) +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))

# group by local_bls > 0.9
plt <- copy(ncorfs_cons)
plt[, grp := ifelse(bls_local > 0.9, 'Local BLS > 0.9', 'local BLS < 0.9')]

p1 <- ggplot(plt[species == 'Human' & method == 'ASR'], aes(x = phylocsf_pc, fill = grp)) +
    geom_density(position = 'identity', alpha = 0.618) +
    coord_cartesian(xlim = c(-50, 25)) +
    scale_fill_brewer(palette = 'Dark2') +
    labs(x = 'PhyloCSF per codon', y = 'Density', fill = NULL, title = 'Human') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))

p2 <- ggplot(plt[species == 'Mouse' & method == 'ASR'], aes(x = phylocsf_pc, fill = grp)) +
    geom_density(position = 'identity', alpha = 0.618) +
    coord_cartesian(xlim = c(-20, 10)) +
    scale_fill_brewer(palette = 'Dark2') +
    labs(x = 'PhyloCSF per codon', y = 'Density', fill = NULL, title = 'Mouse') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))
p1 + p2 + plot_layout(guides = 'collect')
ggsave('figures/ncorf_phylocsf_pc_by_localBLS_ASR.pdf', width = 8, height = 3)

p1 <- ggplot(plt[species == 'Human' & method == 'MRCA'], aes(x = phylocsf_pc, fill = grp)) +
    geom_density(position = 'identity', alpha = 0.618) +
    coord_cartesian(xlim = c(-50, 25)) +
    scale_fill_brewer(palette = 'Dark2') +
    labs(x = 'PhyloCSF per codon', y = 'Density', fill = NULL, title = 'Human') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))

p2 <- ggplot(plt[species == 'Mouse' & method == 'MRCA'], aes(x = phylocsf_pc, fill = grp)) +
    geom_density(position = 'identity', alpha = 0.618) +
    coord_cartesian(xlim = c(-20, 10)) +
    scale_fill_brewer(palette = 'Dark2') +
    labs(x = 'PhyloCSF per codon', y = 'Density', fill = NULL, title = 'Mouse') +
    scale_y_continuous(expand = expansion(mult = c(0, 0.05)))
p1 + p2 + plot_layout(guides = 'collect')
ggsave('figures/ncorf_phylocsf_pc_by_localBLS_MRCA.pdf', width = 8, height = 3)

tmp <- plt[, .(log10p = log10(wilcox.test(
    phylocsf_pc[bls_local > 0.9], phylocsf_pc[bls_local < 0.9]
)$p.value)), keyby = .(species, method)]
# "  species method     log10p
#     <char> <char>      <num>
# 1:   Human    ASR       -Inf
# 2:   Human   MRCA -152.51305
# 3:   Mouse    ASR  -98.74937
# 4:   Mouse   MRCA  -94.34251"

tmp <- plt[, .N, keyby = .(species, method, grp, phylocsf_pc > 0)]
tmp <- dcast(tmp, species + method + grp ~ phylocsf_pc)
tmp[, frac := `TRUE`/(`FALSE` + `TRUE`)]


tmp[, fisher.test(as.matrix(.SD))$p.value, by = .(species, method), .SDcols = c('FALSE', 'TRUE')]
#    species method            V1
#     <char> <char>         <num>
# 1:   Human    ASR 1.597624e-150
# 2:   Human   MRCA  2.768208e-60
# 3:   Mouse    ASR  4.721148e-34
# 4:   Mouse   MRCA  5.361553e-35
```
